mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java

```diff
@@ -0,0 +1,345 @@
+    private Schema expectedSchema;
+    private Schema tableSchema;
+    private ReadSupport<T> readSupport;
+    private Closeable currentCloseable;
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext context) {
+      Configuration conf = context.getConfiguration();
+      CombinedScanTask task = ((IcebergSplit) split).task;
+      this.context = context;
+      this.tasks = task.files().iterator();
+      this.tableSchema = SchemaParser.fromJson(conf.get(TABLE_SCHEMA));
+      String readSchemaStr = conf.get(READ_SCHEMA);
+      if (readSchemaStr != null) {
+        this.expectedSchema = SchemaParser.fromJson(readSchemaStr);
+      }
+      this.readSupport = readSupport(conf);
+      this.currentIterator = open(tasks.next());
+    }
+
+    @Override
+    public boolean nextKeyValue() throws IOException {
+      while (true) {
+        if (currentIterator.hasNext()) {
+          currentRow = currentIterator.next();
+          return true;
+        } else if (tasks.hasNext()) {
+          currentCloseable.close();
+          currentIterator = open(tasks.next());
+        } else {
+          return false;
+        }
+      }
+    }
+
+    @Override
+    public Void getCurrentKey() {
+      return null;
+    }
+
+    @Override
+    public T getCurrentValue() {
+      return currentRow;
+    }
+
+    @Override
+    public float getProgress() {
+      return context.getProgress();
+    }
+
+    @Override
+    public void close() throws IOException {
+      currentCloseable.close();
+    }
+
+    private ReadSupport<T> readSupport(Configuration conf) {
+      String readSupportClassName = conf.get(READ_SUPPORT);
+      try {
+        return DynClasses
+            .builder()
+            .impl(readSupportClassName)
+            .<ReadSupport<T>>buildChecked()
+            .newInstance();
+      } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
+        throw new RuntimeException(String.format("Unable to instantiate read support %s", readSupportClassName), e);
+      }
+    }
+
+    private Iterator<T> open(FileScanTask currentTask) {
+      DataFile file = currentTask.file();
+      // schema of rows returned by readers
+      PartitionSpec spec = currentTask.spec();
+      Set<Integer> idColumns = spec.identitySourceIds();
+
+      boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+      Schema readSchema = expectedSchema != null ? expectedSchema : tableSchema;
+      if (hasJoinedPartitionColumns) {
+        readSchema = TypeUtil.selectNot(tableSchema, idColumns);
+        Schema partitionSchema = TypeUtil.select(tableSchema, idColumns);
+        return Iterators.transform(
+            open(currentTask, readSchema),
+            row -> readSupport.withPartitionColumns(row, partitionSchema, spec, file.partition()));
+      } else {
+        return open(currentTask, readSchema);
+      }
+    }
+
+    private Iterator<T> open(FileScanTask currentTask, Schema readSchema) {
```

Comment:
I read the pig & spark source reader, seems we could abstract the FileScanTask open(..) method with readSchema,  so that all the compute engine can share that part of code,  maybe we can defined it as: 
```
Iterator<T> open(FileScanTask currentTask, Schema readSchema, ReadSupport<T> readSupport).
```
I think it could be a file-level abstraction ..

Reply:
I also have the same feelings, with more framework supported, there could be a bunch of duplicated codes. It would be better to have an abstract layer which could be adapted to all the frameworks.

Reply:
Pig will eventually rely on this InputFormat, but I agree generally, if we can abstract away common code, it would be great.

Reply:
Yeah, there may be a good way to extract common code. For now, let's do that in a separate PR though to keep this focused on the MR support.